### PR TITLE
(Fix 3813) Handle misconfigured Avalara api

### LINKED ISSUE #3813: Avalara does not gracefully handle misconfiguration errors
### Expected behavior
Avalara should gracefully handle bad credentials.

### Actual behavior
An error is thrown and the UI locks up

With APILoginId not configured:
Server error logged:
```
19:24:19.092Z  WARN Reaction: The Avalara package cannot function unless apiLoginId is configured
```

image

```
Exception in delivering result of invoking 'accounts/validateAddress': TypeError: Cannot read property 'validated' of undefined
    at http://localhost:3000/app/app.js?hash=5f81338ecbdeddcb9a19780dd2dc5fca3d8b6cf9:27299:17
    at MethodInvoker._callback (http://localhost:3000/packages/meteor.js?hash=b0f12795c8cc1423b5850502871996903f947ed5:1167:22)
    at MethodInvoker._maybeInvokeCallback (http://localhost:3000/packages/ddp-client.js?hash=69c1d15adcf9b913cb4704b652adeff4bc462aa8:209:14)
    at MethodInvoker.receiveResult (http://localhost:3000/packages/ddp-client.js?hash=69c1d15adcf9b913cb4704b652adeff4bc462aa8:233:12)
    at Connection._livedata_result (http://localhost:3000/packages/ddp-client.js?hash=69c1d15adcf9b913cb4704b652adeff4bc462aa8:1975:11)
    at Connection.onMessage (http://localhost:3000/packages/ddp-client.js?hash=69c1d15adcf9b913cb4704b652adeff4bc462aa8:2169:14)
    at http://localhost:3000/packages/socket-stream-client.js?hash=7063aa833005aff7419b3f8c222e877582f1a3e0:262:13
    at Array.forEach (<anonymous>)
    at ClientStream.forEachCallback (http://localhost:3000/packages/socket-stream-client.js?hash=7063aa833005aff7419b3f8c222e877582f1a3e0:353:33)
    at SockJS.socket.onmessage (http://localhost:3000/packages/socket-stream-client.js?hash=7063aa833005aff7419b3f8c222e877582f1a3e0:261:18)
```

That error spawns from this line `if (res.validated)`
```
AutoForm.hooks({
  addressBookAddForm: {
    onSubmit: function (insertDoc) {
      var that = this;
      this.event.preventDefault();
      var addressBook = $(this.template.firstNode).closest(".address-book");
      Meteor.call("accounts/validateAddress", insertDoc, function (err, res) {
        // if the address is validated OR the address has already been through the validation process, pass it on
=====>  if (res.validated) {
          Meteor.call("accounts/addressBookAdd", insertDoc, function (error, result) {
            if (error) {
              Alerts.toast(i18next.t("addressBookAdd.failedToAddAddress", {
                err: error.message
              }), "error");
```

If the package is not configured correctly, then we see this in the server console
```
modules.js:67965 Objecterr: {message: "The Avalara package is not configured correctly. Cannot continue [bad-configuration]", name: "Error", stack: "Error: The Avalara package is not configured corre…7063aa833005aff7419b3f8c222e877582f1a3e0:2044:15)", code: undefined, signal: undefined}hostname: "localhost:3000"level: 50msg: "Failed to add to cart."name: "reaction-client"pid: undefinedtime: Wed Feb 21 2018 10:17:21 GMT-0800 (PST) {}v: 0__proto__: Object
```

### Steps to reproduce the behavior
1. Enable Avalara with either bad or missing credentials
1. Observe that checkout is blocked

### Acceptance Criteria
Use Avalara with:
- Bad credentials
- No credentials
- Good credentials

### Versions
Node: 8.9.0
NPM: 5.6.0
Meteor Node: 8.9.4
Meteor NPM: 5.6.0
Reaction CLI: 0.27.0
Reaction: 1.8.0
Reaction branch: release-1.8.0
Docker: 17.12.0-ce

For enhancements or other issues, please describe, in as much detail as
possible, how the app should be changed, and how this would benefit other users and developers.

### PR DESCRIPTION
**Pull request by zenweasel**
Resolves #3813
Impact: **minor**  
Type: **bugfix**

## Issue
In order to make sure that the site operator is made aware of a misconfigured API it used to throw errors and basically stop operation when the API is misconfigured. Now we just log errors and continue.

## Solution
Don't block order processing. Allow orders to be placed and just log errors

## Breaking changes
None

## Testing
1. Enable Avalara and remove the apiLoginId
1. Place an item in cart and checkout
1. Observe that you can place an order even though tax calculation is not working
1. Observe that server errors are thrown to the console which is correct

This does not fix the issue with address validation when the API is configured correctly.

### CODE REVIEW

**Comment by zenweasel** [comment]
Making WIP because I will also fix #3518 in this PR

**Comment by aaronjudd** [comment]
@zenweasel why did you expand the scope of this PR? Could we take this and follow up with PR for #3518?

**Comment by zenweasel** [comment]
@aaronjudd Yes, I will fix up today and remove WIP. I thought I was going to follow up quickly with the rest but it didn't turn out that way.

File: imports/plugins/included/taxes-avalara/server/hooks/hooks.js [hunk 1]
```diff
@@ -32,8 +33,9 @@ MethodHooks.after("taxes/calculate", (options) => {
         const taxAmount = taxes.reduce((totalTaxes, tax) => totalTaxes + tax.tax, 0);
         const taxRate = taxAmount / taxCalc.calcTaxable(cartToCalc);
         Meteor.call("taxes/setRate", cartId, taxRate, taxes);
-      } else if (result.error.errorCode === 503) {
-        Logger.error("timeout error: do nothing here");
+        // for bad auth, timeout, or misconfiguration there's nothing we can do so keep moving
+      } else if (_.includes([503, 400, 401], result.error.errorCode)) {
```
Comment (jshimko) [inline imports/plugins/included/taxes-avalara/server/hooks/hooks.js]: We could probably skip `lodash` here.

This...

```js
_.includes([503, 400, 401], result.error.errorCode)
```

Could be...

```js
[503, 400, 401].includes(result.error.errorCode)
```

File: imports/plugins/included/taxes-avalara/server/methods/taxCalc.js [hunk 2]
```diff
@@ -292,6 +324,18 @@ taxCalc.validateAddress = function (address) {
...
+    if (result.error.type === "apiError") {
+      // If we have a problem with the API there's no reason to tell the customer
+      // so let's consider this unvalidated but move along
+      Logger.info("API error, ignoring address validation");
+    }
+
+    if (result.error.type === "validationError") {
+      // We received a validation error so we need somehow pass this up to the client
+      Logger.info("Address Validation Error");
+    }
```
Comment (jshimko) [inline imports/plugins/included/taxes-avalara/server/methods/taxCalc.js]: These two loggers should probably be a `warn` or `error`, yes?

File: imports/plugins/included/taxes-avalara/server/methods/taxCalc.js [hunk 3]
```diff
@@ -180,6 +200,18 @@ function avaGet(requestUrl, options = {}, testCredentials = true) {
...
+        errorCode: 400,
+        type: "apiFailure",
+        errorDetails: {
+          message: "API is not configured"
+        }
+      }
+    };
+  }
   const appVersion = Reaction.getAppVersion();
   const meteorVersion = _.split(Meteor.release, "@")[1];
```
Comment (jshimko) [inline imports/plugins/included/taxes-avalara/server/methods/taxCalc.js]: I know this isn't your code, but while we're here...

This is an unnecessary usage of `lodash` for a basic string manipulation.
```js
// nope
_.split(Meteor.release, "@")[1];

// yep
Meteor.release.split("@")[1]
```

**Comment by zenweasel** [comment]
@jshimko is there a reason you can't make these changes and approve? They seem like solid points

**Comment by jshimko** [comment]
No, there isn't.  But I wasn't aware that was the workflow.

**Comment by zenweasel** [comment]
@jshimko Changes made as requested

**Pull request merged**